controller/volume_controller.go

```diff
@@ -2367,52 +2372,49 @@ func (vc *VolumeController) processMigration(v *longhorn.Volume, es map[string]*
 		return nil
 	}
 
-	if len(es) > 2 {
-		return fmt.Errorf("BUG: volume %v: more than two engines exists", v.Name)
-	}
-
+	// the only time there should be more then 1 engines is when we are migrating or upgrading
+	// if there are more then 1 and we no longer have a migration id set we can cleanup the extra engine
 	if !vc.isVolumeMigrating(v) {
-		if len(es) != 2 {
+		if len(es) < 2 {
```

Comment:
What if the extra engines are cleaned up but the volume update is failed?

Reply:
Check you check this comment? If line 2388 `vc.getCurrentEngineAndCleanupOthers(v, es)` succeeded but volume status update failed due to the conflict, there is no way to redo line 2883

Reply:
Needs to be handled outside of the migration, since it's the same case as v.Spec.NodeID != v.Status.CurrentNodeID
i.e. share-manager detach/attach issue.

Updated the PR with a fixup commit, we need to prioritize a refactor of the volume controller to address this tech debt.

Reply:
But the case is different from the share manager part. 
The engine is already on `v.Spec.NodeID` in this case. Longhorn doesn't need to detach the volume then reattach it to `v.Spec.NodeID`

Reply:
See https://github.com/longhorn/longhorn-manager/pull/822/commits/03a7e57a39da2545d015c0980a5f20f35dc3f572

Reply:
OK, I see.